core/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -0,0 +1,610 @@
+import org.elasticsearch.common.regex.Regex;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.indices.IndexClosedException;
+import org.elasticsearch.indices.IndexMissingException;
+
+import java.util.*;
+
+import static com.google.common.collect.Lists.newArrayList;
+import static com.google.common.collect.Maps.newHashMap;
+
+public class IndexNameExpressionResolver {
+
+    private final ImmutableList<IndexExpressionFilter> parsers;
+
+    @Inject
+    public IndexNameExpressionResolver() {
+        parsers = ImmutableList.<IndexExpressionFilter>of(new WildcardExpressionFilter());
+    }
+
+    public String[] concreteIndices(ClusterState state, IndicesRequest request) throws IndexMissingException, IllegalArgumentException {
+        Context context = new Context(state, request.indicesOptions());
+        return concreteIndices(context, request.indices());
+    }
+
+    public String[] concreteIndices(ClusterState state, IndicesOptions options, String... indexExpressions) throws IndexMissingException, IllegalArgumentException {
+        Context context = new Context(state, options);
+        return concreteIndices(context, indexExpressions);
+    }
+
+    /**
+     * Translates the provided index expression into actual concrete indices.
+     *
+     * @param indexExpressions expressions that can be resolved to alias or index names.
+     * @param context   how the aliases or indices need to be resolved to concrete indices
+     * @return the obtained concrete indices
+     * @throws IndexMissingException if one of the aliases or indices is missing and the provided indices options in the context
+     * don't allow such a case, or if the final result of the indices resolution is no indices and the indices options
+     * in the context don't allow such a case.
+     * @throws IllegalArgumentException if one of the aliases resolve to multiple indices and the provided
+     * indices options in the context don't allow such a case.
+     */
+    public String[] concreteIndices(Context context, String... indexExpressions) throws IndexMissingException, IllegalArgumentException {
+        if (indexExpressions == null || indexExpressions.length == 0) {
+            indexExpressions = new String[]{MetaData.ALL};
+        }
+        MetaData metaData = context.getState().metaData();
+        IndicesOptions options = context.getOptions();
+        boolean failClosed = options.forbidClosedIndices() && options.ignoreUnavailable() == false;
+        boolean failNoIndices = options.ignoreUnavailable() == false;
+
+        // TODO: this exception in the logic must be revised:
+        // I think this is an inconsistency (bug) in the old single expressions logic. (there was different logic for a single and multiple index logic)
+        // This must be addressed in a follow up issue
+        if (indexExpressions.length == 1) {
+            failNoIndices = options.allowNoIndices() == false;
+        }
```

Comment:
I did quite some digging around this special case. It is there from the early days of indices options and back when I added unit tests for it to MetaDataTests I probably just added tests to verify the current behaviour (assuming that what we had was ok). Looking at it more I don't get this test that fails when you remove this special case:

```
IndicesOptions noExpandErrorUnavailable = IndicesOptions.fromOptions(false, true, false, false);

            String[] results = md.concreteIndices(noExpandErrorUnavailable, "baz*");
            assertThat(results, emptyArray());

            try {
                md.concreteIndices(noExpandErrorUnavailable, "foo", "baz*");
                fail();
            } catch (IndexMissingException e) {
                assertThat(e.index().name(), equalTo("baz*"));
            }
```

it makes no sense to me that with same options, a different number of indices leads to a different result... I think we could adapt the test and remove this special case (in a separate issue it's fine too)